### PR TITLE
London | 26-ITP-Jan | Angela McLeary | Sprint 1 | Sprint 1

### PR DESCRIPTION
**Pull request by AngelaMcLeary**
## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

This PR is about fixing median.js, Implementing dedupe.js, max.js, sum.js and refactoring include.js.

### CODE REVIEW

File: Sprint-1/fix/median.js [hunk 1]
```diff
@@ -6,9 +6,35 @@
...
 
 function calculateMedian(list) {
-  const middleIndex = Math.floor(list.length / 2);
-  const median = list.splice(middleIndex, 1)[0];
-  return median;
-}
-
+  if (!Array.isArray(list) || list.length === 0) return null;
+  //If the items are not an array or if the array is empty return null
+  const numbers = list.filter((num) => typeof num === "number");
```
Comment (cjyuan) [inline Sprint-1/fix/median.js]: Do you plan to consider also `-Infinity`, `Infinity`, and `NaN` in the median calculation (and in the functions in `implement/max.js` and `implement/sum.js`)?

File: Sprint-1/fix/median.js [hunk 2]
```diff
@@ -6,9 +6,35 @@
...
+  if (numbers.length === 0) return null;
+  //checks the array before sorting and removing anything that is not a valid item ie not a number.
+  const sortedList = [...numbers].sort((a, b) => a - b);
```
Comment (cjyuan) [inline Sprint-1/fix/median.js:29]: Is it necessary to create a copy of `numbers` first before sorting in this function?

File: Sprint-1/implement/dedupe.test.js [hunk 3]
```diff
@@ -16,13 +16,27 @@ E.g. dedupe([1, 2, 1]) returns [1, 2]
...
+  expect(dedupe([])).toEqual([]);
+});
 
 // Given an array with no duplicates
 // When passed to the dedupe function
 // Then it should return a copy of the original array
+test("given an array has no duplicates, it returns a copy of the original array", () => {
+  expect(dedupe([1, 2, 3])).toEqual([1, 2, 3]);
+  expect(dedupe([5, 1, 4])).toEqual([5, 1, 4]);
+});
```
Comment (cjyuan) [inline Sprint-1/implement/dedupe.test.js:31]: This test should fail if the function returns the original array (instead of a **copy** of the original array).

The current test checks only if both the original array and the returned array contain identical elements. 
In order to validate the returned array is a different array, we need an **additional** check.

Can you find out what this additional check is?

File: Sprint-1/implement/max.test.js [hunk 4]
```diff
@@ -16,28 +16,51 @@ const findMax = require("./max.js");
...
+});
+
 // Given an array with non-number values
 // When passed to the max function
 // Then it should return the max and ignore non-numeric values
 
+test("ignores non-number values and returns the max number", () => {
+  expect(findMax(["Not", "A", "Number", 75, 85, 105])).toEqual(105);
+});
+
```
Comment (cjyuan) [inline Sprint-1/implement/max.test.js:60]: When a string representing a valid numeric literal (for example, `"300"`) is compared to a number, 
JavaScript first converts the string into its numeric equivalent before performing the comparison. 
As a result, the expression `20 < "300"` evaluates to `true`.

To test if the function can correctly ignore non-numeric values,
consider including a string such as `"300"` in the relevant test cases.

File: Sprint-1/implement/sum.test.js [hunk 5]
```diff
@@ -13,24 +13,40 @@ const sum = require("./sum.js");
...
+  expect(sum([5, 2, -3])).toEqual(4);
+});
 
 // Given an array with decimal/float numbers
 // When passed to the sum function
 // Then it should return the correct total sum
-
+test("given an array with decimal float numbers, it should return the total sum", () => {
+  expect(sum([1.5, 2.5, 3.5])).toEqual(7.5);
+});
```
Comment (cjyuan) [inline Sprint-1/implement/sum.test.js:39]: Decimal numbers in most programming languages (including JS) are internally represented in "floating point number" format. Floating point arithmetic is not exact. For example, the result of `46.5678 - 46 === 0.5678` is false because `46.5678 - 46` only yield a value that is very close to `0.5678`. Even changing the order in which the program add/subtract numbers can yield different values.

So the following could happen
```javascript
  expect( 1.2 + 0.6 + 0.005 ).toEqual( 1.805 );                // This fail
  expect( 1.2 + 0.6 + 0.005 ).toEqual( 1.8049999999999997 );   // This pass
  expect( 0.005 + 0.6 + 1.2 ).toEqual( 1.8049999999999997 );   // This fail

  console.log(1.2 + 0.6 + 0.005 == 1.805);  // false
  console.log(1.2 + 0.6 + 0.005 == 0.005 + 0.6 + 1.2); // false
```

**Can you find a more appropriate way to test a value (that involves decimal number calculations) for equality?**

Suggestion: Look up 
- Checking equality in floating point arithmetic in JavaScript 
- Checking equality in floating point arithmetic with Jest

File: Sprint-1/fix/median.js (hunk 2 above)
Comment (AngelaMcLeary) [reply Sprint-1/fix/median.js:29]: Hi @cjyuan, It is not strictly necessary, but it is good practice to do this to avoid modifying the original array and keeping it in tact.

File: Sprint-1/fix/median.js (hunk 1 above)
Comment (AngelaMcLeary) [reply Sprint-1/fix/median.js]: Hi @cjyuan, Thank you for your feedback. I intend to. I have updated the code to accordingly. It should now pass the test correctly.

File: Sprint-1/implement/dedupe.test.js (hunk 3 above)
Comment (AngelaMcLeary) [reply Sprint-1/implement/dedupe.test.js:31]: Hi @cjyuan, Thank you for the feedback. I added an additional assertion to ensure the dedupe function returns a new array instance, not the original array. The previous test only checked for value equality using .toEqual(), which would still pass even if the function simply returned the input array.To properly validate the acceptance criteria, the test is now fortified.

File: Sprint-1/implement/max.test.js (hunk 4 above)
Comment (AngelaMcLeary) [reply Sprint-1/implement/max.test.js:60]: Hi @cjyuan, Thank you for your feedback. I have updated the test to include an edge case for numerical string.

File: Sprint-1/implement/sum.test.js (hunk 5 above)
Comment (AngelaMcLeary) [reply Sprint-1/implement/sum.test.js:39]: Hi @cjyuan, Thank you for your feedback. I understand now why decimal values need special handling. toBeCloseTo() is a better choice because it checks that the result is close enough within a small margin, which avoids issues caused by floating‑point rounding.

File: Sprint-1/implement/dedupe.test.js [hunk 6]
```diff
@@ -16,13 +16,29 @@ E.g. dedupe([1, 2, 1]) returns [1, 2]
...
 
 // Given an array with no duplicates
 // When passed to the dedupe function
 // Then it should return a copy of the original array
+test("given an array has no duplicates, it returns a copy of the original array", () => {
+ const originalArray = [1, 2, 3];
+ const result = dedupe(originalArray);
+  expect(result).toEqual(originalArray);
+  expect(result).not.toBe(originalArray);
```
Comment (cjyuan) [inline Sprint-1/implement/dedupe.test.js:30]: This test can verify the function can return a different array.

However, there is a chance that, even though `result` has incorrect elements (for example, `[1, 1, 1]`), 
the test on line 29 could still pass. Can you figure out why, and then improve the test accordingly?

**Review by cjyuan — COMMENTED** [review]
Other changes are good. Just one issue left.

File: Sprint-1/implement/sum.test.js (hunk 5 above)
Comment (cjyuan) [reply Sprint-1/implement/sum.test.js:39]: Note: The default value of the second parameter of `toBeCloseTo()` is only 2   
See: https://jestjs.io/docs/expect#tobeclosetonumber-numdigits

For the default value 2, the test criterion is `Math.abs(expected - received) < 0.005`, which may not be precise enough in some applications.

File: Sprint-1/implement/dedupe.test.js (hunk 6 above)
Comment (AngelaMcLeary) [reply Sprint-1/implement/dedupe.test.js:30]: Hi @cjyuan, Thanks for your feedback. I have done some research and understand that this only checks value equality, so it could still pass even if the function returns the same array or mutates it. Added checks to make sure a new array is returned and the original input isn't changed. I have added additional test.

**Comment by cjyuan** [comment]
Changes look good. Well done.

**Comment by illicitonion** [comment]
Closing PR because the January ITP run has finished. Feel free to re-open if you're still working on it.
